codegenerator/cli/templates/static/codegen/src/eventFetching/SourceManager.res

```diff
@@ -0,0 +1,176 @@
+  logger: Pino.t,
+  maxPartitionConcurrency: int,
+  mutable isWaitingForNewBlock: bool,
+  mutable allPartitionsFetchingState: array<partitionFetchingState>,
+  // Don't use fetchingPartitions size, but have a separate counter
+  // to take into consideration partitions fetching for previous states (before rollback)
+  mutable fetchingPartitionsCount: int,
+  // Keep track on the current state id
+  // to work with correct state during rollbacks & preRegistration
+  mutable currentStateId: int,
+}
+
+let make = (~maxPartitionConcurrency, ~logger) => {
+  logger,
+  maxPartitionConcurrency,
+  isWaitingForNewBlock: false,
+  // Don't prefill with empty partitionFetchingState,
+  // since partitions might be added with the lifetime of the application.
+  // So lazily create fetchingState, when we execute a new partition
+  allPartitionsFetchingState: [],
+  currentStateId: 0,
+  fetchingPartitionsCount: 0,
+}
+
+exception FromBlockIsHigherThanToBlock({fromBlock: int, toBlock: int})
+
+let fetchBatch = async (
+  sourceManager: t,
+  ~allPartitions: PartitionedFetchState.allPartitions,
+  ~currentBlockHeight,
+  ~executePartitionQuery,
+  ~waitForNewBlock,
+  ~onNewBlock,
+  ~maxPerChainQueueSize,
+  ~setMergedPartitions,
+  ~stateId,
+) => {
+  if stateId < sourceManager.currentStateId {
+    ()
+  } else {
+    if stateId != sourceManager.currentStateId {
+      sourceManager.currentStateId = stateId
+
+      // Reset instead of clear, so updating state from partitions from prev state doesn't corrupt data
+      sourceManager.allPartitionsFetchingState = []
+    }
+    let {logger, allPartitionsFetchingState, maxPartitionConcurrency} = sourceManager
+
+    let fetchingPartitions = Utils.Set.make()
+    // Js.Array2.forEachi automatically skips empty items
+    allPartitionsFetchingState->Js.Array2.forEachi(({isFetching}, partitionId) => {
+      if isFetching {
+        let _ = fetchingPartitions->Utils.Set.add(partitionId)
+      }
+    })
+    let readyPartitions =
+      allPartitions->PartitionedFetchState.getReadyPartitions(
+        ~maxPerChainQueueSize,
+        ~fetchingPartitions,
+      )
+
+    let mergedPartitions = Js.Dict.empty()
+    let hasQueryWaitingForNewBlock = ref(false)
+    let queries = readyPartitions->Array.keepMap(({fetchState, partitionId}) => {
+      let mergedFetchState = fetchState->FetchState.mergeRegistersBeforeNextQuery
+      if mergedFetchState !== fetchState {
+        mergedPartitions->Js.Dict.set(partitionId->(Utils.magic: int => string), mergedFetchState)
+      }
+      switch mergedFetchState->FetchState.getNextQuery(~partitionId) {
+      | Done => None
+      | NextQuery(nextQuery) =>
+        switch allPartitionsFetchingState->Belt.Array.get(partitionId) {
+        // Deduplicate queries when fetchBatch is called after
+        // isFetching was set to false, but state isn't updated with fetched data
+        | Some({lastFetchedQueryId}) if lastFetchedQueryId === toQueryId(nextQuery) => None
```

Comment:
Out of interest does `===` vs `==` have any effect on strings?

Reply:
No, it's the same generated code, I just don't like using `==` both in Js and ReScript 😅